File: Pharmaceutical_Company_Example/02_Structural/R/Adapter.R
# Adaptee
ClinicalTrialSystem <- setRefClass(
  "ClinicalTrialSystem",
  fields = list(
    trial_data = "character"
  ),
  methods = list(
    initialize = function(drug_name) {
      .self$trial_data <- paste(drug_name, "clinical trial data")
    },
    get_trial_data = function() {
      return(.self$trial_data)
    }
  )
)

# Target
RegulatoryAPI <- setRefClass(
  "RegulatoryAPI",
  methods = list(
    upload_data_to_regulator = function(data) {
      message("Uploading the following data to regulatory authorities: ", data)
    }
  )
)

# Adapter to bridge between ClinicalTrialSystem (Adaptee) and RegulatoryAPI (Target)
RegulatoryAdapter <- setRefClass(
  "RegulatoryAdapter",
  fields = list(
    clinical_trial_system = "ANY",
    regulatory_api = "ANY"
  ),
  methods = list(
    initialize = function(clinical_trial_system, regulatory_api) {
      .self$clinical_trial_system <- clinical_trial_system
      .self$regulatory_api <- regulatory_api
    },
    submit_trial_data = function() {
      trial_data <- .self$clinical_trial_system$get_trial_data()
      .self$regulatory_api$upload_data_to_regulator(trial_data)
    }
  )
)


clinical_system_1 <- ClinicalTrialSystem$new("Test Small Molecule 01")
clinical_system_2 <- ClinicalTrialSystem$new("Test Small Molecule 02")
regulatory_api <- RegulatoryAPI$new()

adapter_1 <- RegulatoryAdapter$new(clinical_system_1, regulatory_api)
adapter_1$submit_trial_data()

adapter_2 <- RegulatoryAdapter$new(clinical_system_2, regulatory_api)
adapter_2$submit_trial_data()

File: Pharmaceutical_Company_Example/02_Structural/JavaScript/Adapter.js
// Adaptee: ClinicalTrialSystem
class ClinicalTrialSystem {
  constructor(drugName) {
    this.trialData = `${drugName} clinical trial data`;
  }

  getTrialData() {
    return this.trialData;
  }
}

// Target: RegulatoryAPI
class RegulatoryAPI {
  uploadDataToRegulator(data) {
    console.log(`Uploading the following data to regulatory authorities: ${data}`);
  }
}

// Adapter to bridge between ClinicalTrialSystem (Adaptee) and RegulatoryAPI (Target)
class RegulatoryAdapter {
  constructor(clinicalTrialSystem, regulatoryAPI) {
    this.clinicalTrialSystem = clinicalTrialSystem;
    this.regulatoryAPI = regulatoryAPI;
  }

  submitTrialData() {
    const trialData = this.clinicalTrialSystem.getTrialData();
    this.regulatoryAPI.uploadDataToRegulator(trialData);
  }
}

// Create instances of ClinicalTrialSystem and RegulatoryAPI
const clinicalSystem1 = new ClinicalTrialSystem("Test Small Molecule 01");
const clinicalSystem2 = new ClinicalTrialSystem("Test Small Molecule 02");
const regulatoryAPI = new RegulatoryAPI();

// Create adapters for both clinical systems
const adapter1 = new RegulatoryAdapter(clinicalSystem1, regulatoryAPI);
adapter1.submitTrialData();

const adapter2 = new RegulatoryAdapter(clinicalSystem2, regulatoryAPI);
adapter2.submitTrialData();
